### ISSUE TITLE
Show loading spinner before each emit

### ISSUE THREAD

**Issue by ThRintelen** [rx-angular/rx-angular#394]
I like the suspense to show loading spinner on startup.

But in this example, i cant show a loading spinner:
I have a table. First comes the loading spinner, then the data will be emit. Now I am searching for something in the table (remote) and the data is reloaded and emit new values.

A loading spinner would also be good in this step.

It would be nice to show a template before each emit.

**Comment by BioPhoton** [comment]
> It would be nice to show a template before each emit.

So atm we assume a reset is required if a new observable was assigned to rxLet.
What I can thing of atm is a e.g. undefined sent to the directive to reset it?

I see such situations and their ergonomic handling as a important thing and we will have to work on a couple POCs to compare them and see how we can get such behaviour in the directive.

Wher we are done with the big refactoring under the hood I will have to create a RFC for that feature and do some research on how it could look. 

If you are interested in that process and want to support, let me know and we can start to slowly work on an RFC and some demos to find a nice solution.

**Comment by ThRintelen** [comment]
Yes, im interested. Let me know if you want to start.

**Comment by BioPhoton** [comment]
So first of all we need to clarify the problem.

As far as I can tell it is "resetting to suspense template"
Here we have to document the problem with a short snippet or so. 

Prior Art:
We should mention that at the moment it is only possible to do it by reassigning a new observable.
We can also take rx-querry for an example usecase (if we do that we can directly create a demo in the demos ap of this repo).  where we use different observables for the template state. (ngSwitch)

We can think of a way to reset it within the same observable. This means the observable should e.g. emit something special. 
This should be explained in the text and as a quick code scatch.

Than we can see if we get feedback. 

An example is the rxFor RFC (#304). But it is in WIP state so it will most probably change. Maybe take the select RFC (#263) as a better example.

We will have to keep in mind that there are rxIf, rxFor, rxSwitch coming and this should also be considered in the RFC.


Did that explain the idea? Just let me know what's unclear.

**Comment by BioPhoton** [comment]
Another good example we just created is #399

**Comment by ThRintelen** [comment]
okay, this is my first RFC, but i will try it :-)

# Problem Solved By The Feature

For rxLet suspense can show e.g. a loading spinner as custom template before the first value emitted.
But what, if the observable emit more values?

It would be nice if i can show another custom template before each next emit.

## Example (table)

In my example i habe a table. The table load initial data (suspense).
By using a remote filter, the data requested again and the next value emit.
In this case, i want to show an other loading spinner.

# Solution

Rename suspense to beforeFirstValue and add new beforeNextValues for speaking names.

```javascript
<ng-container
  *rxLet="
    observableNumber$;
    let n;
    error: error;
    complete: complete;
    beforeFirstValue: beforeFirstValue;
    beforeNextValues: beforeNextValues;
  "
>
  <app-number [number]="n"></app-number>
</ng-container>

<ng-template #error>ERROR</ng-template>
<ng-template #complete>COMPLETE</ng-template>
<ng-template #beforeFirstValue>BEFORE FIRST VALUE EMITTED</ng-template>
<ng-template #beforeNextValues>BEFORE NEXT VALUE EMITTED</ng-template>
```

# Additional Context

Also think about it in upcoming fetures

- [ ] rxIf
- [ ] rxFor
- [ ] rxSwitch

**Comment by BioPhoton** [comment]
Pretty nice!! 
I may help a little bit to polish it up. :)

> It would be nice if I can show another custom template before each next emit.

Here I would need more information regarding when before is. Basically, as soon as the value arrives it's before the next emit. You see what I mean?

I can imagine the moment for "before the next emit" has a trigger. So we should somehow involve that trigger in the solution. 

If this is not the case, please elaborate a bit more on the "before emit" situation.

**Comment by ThRintelen** [comment]
Thank you :-)
I think i know what you mean. I have updated my RFC. Is that going in the right direction?

**Comment by BioPhoton** [comment]
I introduced template triggers for one of the stable releases. You can test it here: https://github.com/rx-angular/rx-angular/blob/master/apps/demos/src/app/features/experiments/structural-directives/rx-let-poc/rx-let-poc.component.ts#L33

if u like and give feedback

**Comment by ThRintelen** [comment]
Great job, i like the idea with the suspenseTrg. It looks like a nice solution for my problem / need.

**Comment by BioPhoton** [comment]
closed as better documented in #450
